src/driver-gskill.c

```diff
@@ -0,0 +1,1365 @@
+					       HID_REQ_GET_REPORT);
+		/*
+		 * Sometimes the mouse just doesn't send anything when it wants
+		 * to tell us it's ready
+		 */
+		if (rc == 0) {
+			continue;
+		} else if (rc < GSKILL_REPORT_SIZE_CMD) {
+			log_error(device->ratbag,
+				  "Error while getting command response from mouse: %d\n",
+				  rc);
+
+			return rc;
+		}
+
+		/* Check the command status bit */
+		switch (buf[1]) {
+		case 0: /* sometimes the mouse gets lazy and just returns a
+			   blank buffer on success */
+		case GSKILL_CMD_SUCCESS:
+			return 0;
+
+		case GSKILL_CMD_IN_PROGRESS:
+			break;
+
+		case GSKILL_CMD_IDLE:
+			log_error(device->ratbag,
+				  "Command response indicates idle status? Uh huh.\n");
+			return GSKILL_CMD_IDLE;
+
+		case GSKILL_CMD_FAILURE:
+			log_error(device->ratbag, "Command failed\n");
+			return GSKILL_CMD_FAILURE;
+
+		default:
+			log_error(device->ratbag,
+				  "Received unknown command status from mouse: 0x%x\n",
+				  buf[1]);
+			return -1;
+		}
+	}
+
+	log_error(device->ratbag,
+		  "Failed to get command response from mouse after %d tries, giving up\n",
+		  retries);
+	return -1;
+}
+
+static int
+gskill_get_active_profile_idx(struct ratbag_device *device)
+{
+	uint8_t buf[GSKILL_REPORT_SIZE_CMD] = { 0x0c, 0xc4, 0x7, 0x0, 0x1 };
+	int rc;
+
+	rc = gskill_general_cmd(device, buf);
+	if (rc) {
+		log_error(device->ratbag,
+			  "Error while getting active profile number from mouse: %d\n",
+			  rc);
+		return -1;
+	}
+
+	return buf[3];
+}
+
+static int
+gskill_set_active_profile(struct ratbag_device *device, unsigned index)
+{
+	uint8_t buf[GSKILL_REPORT_SIZE_CMD] = { 0x0c, 0xc4, 0x7, index, 0x0 };
+	int rc;
+
+	rc = gskill_general_cmd(device, buf);
+	if (rc) {
+		log_error(device->ratbag,
+			  "Error while changing active profile on mouse: %d\n",
+			  rc);
+		return -1;
+	}
+
+	return 0;
+}
+
+static int
+gskill_get_profile_count(struct ratbag_device *device)
+{
+	uint8_t buf[GSKILL_REPORT_SIZE_CMD] = { 0x0c, 0xc4, 0x12, 0x0, 0x1 };
+	int rc;
+
+	rc = gskill_general_cmd(device, buf);
+	if (rc) {
+		log_error(device->ratbag,
+			  "Error while getting the number of profiles: %d\n",
+			  rc);
+		return -1;
+	}
+
+	return buf[3];
+}
+
+static int
+gskill_set_profile_count(struct ratbag_device *device, unsigned int count)
+{
+	uint8_t buf[GSKILL_REPORT_SIZE_CMD] = { 0x0c, 0xc4, 0x12, count, 0x0 };
+	int rc;
+
+	rc = gskill_general_cmd(device, buf);
+	if (rc) {
+		log_error(device->ratbag,
+			  "Error while setting the number of profiles: %d\n",
+			  rc);
+		return -1;
+	}
+
+	return 0;
+}
+
+/*
+ * This is used for setting the profile index argument on the mouse for both
+ * reading and writing profiles
+ */
+static int
+gskill_select_profile(struct ratbag_device *device, unsigned index, bool write)
+{
+	uint8_t buf[GSKILL_REPORT_SIZE_CMD] = { 0x0c, 0xc4, 0x0c, index, write };
+	int rc;
+
+	/* Indicate which profile we want to retrieve */
+	rc = ratbag_hidraw_raw_request(device, GSKILL_GENERAL_CMD,
```

Comment:
Unfortunately the protocol for this thing is kind of all over the place. While we use the same command tag (`GSKILL_GENERAL_CMD`) here to select the profile, the mouse gets very unhappy if we try to follow the procedure used for normal commands here as well (e.g. set report followed by wait for mouse to be ready).